### ISSUE TITLE
Unable to use datepicker with an external update query

### ISSUE THREAD

**Issue by mjashanks** [Budibase/budibase#1504]
**Describe the bug**
Originally raised here: https://github.com/Budibase/budibase/discussions/1500

I am trying to use a Budibase datepicker to update a postgres date column.

**My Setup**
- Postgres date column called `created`
- Query, that takes a parameter of `created` and runs and `update` 
- In design, a `Form`, `Field Group`, `DatePicker`. The datepicker's `Field` property is `created` 
- A save button, which executes the query, and passes `{{ Form.Fields.created }}`. Have also tried `{{ date Form.Fields.created  "YYYY-MM-DD" }}` and various other bindings

The request that gets sent by `Execute Query` sends
- `[Object object]` for `{{ Form.Fields.created }}`. 
- `Invalid Date` for `{{ date Form.Fields.created  "YYYY-MM-DD" }}`

**Comment by aptkingston** [comment]
The value of actual date picker (thanks to Flatpickr) is a full JS date object. Not sure how to handle that in Handlebars, but it's not a string so it might not be trivial. We could maybe do something similar to what happens with relationships, where if you bind to a relationship field it sneakily replaces it behind the scenes with the length of the array, rather than the array itself. We could use this same idea and replace date bindings with either the UTC string or the timestamp, making it easy to work with.

**Comment by shogunpurple** [comment]
Bump

**Comment by mjashanks** [comment]
I still have an issue with this. The date picker is timezone aware - I don't think it should be right now.

For example. I set my timezone to UTC -5:00 choose the date `2nd November 2021 17:00`. When I click save, the request sends the date `2nd November 2021 12:00`.

I think it should send `2nd November 2021 17:00` - i.e. we want the ISO format, but no conversion to UTC.

**Comment by aptkingston** [comment]
But if you pick `2nd November 2021 12:00` in your time zone, and then we send `2021-11-02T12:00:00.000Z` then we are not sending what you picked, we're sending a time 5 hours in the past. Reading that date back into Budibase will show it as 5 hours ago, while if we send the correct `2021-11-02T17:00:00.000Z` and read it back into Budibase then it should show exactly the time you picked of `2nd November 2021 12:00`.

What's the issue you're having with it at the moment?

**Comment by mjashanks** [comment]
So - I can see what you mean here. It works because it gets correctly converted on the way in and on the way out. What we are doing is the most correct thing to do.

The issue comes from here (watch the video) : https://github.com/Budibase/budibase/discussions/3147#discussioncomment-1570882

The expectation from the user is that the date will appear in the database with the exact numbers that they entered. However, I just noticed that they are using a `timestamp`, which always gets stored in UTC anyway.

However, If they were to change the column to a `DATETIME`, then this would still be stored as UTC, as the MySQL `DATETIME` type does not store a timezone, and this is how we are sending it.

So, the issue is that the user wants to store a date exactly how they entered it without a timezone. What they want is `DATETIME` support - our datepicker only supports `TIMESTAMP`.

Is it best to have an option on the Date Picker - `Timestamp` or `Timezone Aware`, ticked by default?

**Comment by aptkingston** [comment]
This looks like a MySQL specific thing. Taken from the docs:

> MySQL converts TIMESTAMP values from the current time zone to UTC for storage, and back from UTC to the current time zone for retrieval. (This does not occur for other types such as DATETIME.)

I wonder if using a TIMESTAMP causes a double conversion, where it receives our value (already in UTC) and doubly converts it again into UTC (although it should be converting it back again like it says). That would explain why it's still wrong for him in that thread. If that's the case, using a DATETIME as his MySQL datatype should not have this affect.

I hate date formats so much. Unix timestamps (milliseconds since epoch) is just the superior absolute reference format and it's just a simple number. But I doubt we can just switch over to using that since we need to support so many DB options.

**Comment by KernOil** [comment]
The issue happens with both SQL formats DATETIME and TIMESTAMP

**Comment by KernOil** [comment]
HI @aptkingston thank you very much for looking into this problem I am having.  were you able to find a potential solution to the SQL Date reading error?

**Comment by mjashanks** [comment]
I think this is still a problem. I know that it's technically correct - if everything is stored in UTC, then it gets converted correctly both ways.

However, this is often not the reality of how people store dates in their databases. The create a `timestamp without timezone` column, and they store dates in their own timezone (without a timezone stamp). Bubibase then assumes that it was saved in UTC.

This was raised again here: https://github.com/Budibase/budibase/discussions/3547 , and can see how it would be common.

I think we need an option on the Date Picker - `Timestamp` or `Timezone Aware` - which determines whether it should convert to/from UTC or not?

I am happy to setup a demo fo this sometime - but I'm going to reopen this, as I think it's a real world problem (even though our implementation is "correct")

**Comment by aptkingston** [comment]
Yea I understand what you mean. I think it would nearly need to be part of the schema though, since other places than just date pickers need to be aware of how to parse and display the date.

A demo would be really helpful to clarify the behaviour and test that a solution works.

**Comment by ggs331** [comment]
In my case, Budibase and MySQL timezone is my local timezone CST(UTC +8), and I have many MySQL fields with date type `DATE`, and this kind of fields didn't store timezone info but just something like `YYYY-MM-DD`. When Budibase query from MySQL, if there's no timezone info then Budibase will assume it's UTC, then it will convert to my local timezone CST which will be plus 8hrs. If I use a repeater, the workaround is use `{{date value "YYYY-MM-DD" "UTC"}}` to force convert it to UTC. But in many case, I use update form to bind the date, and there's no way to make the convert. It brings many trouble. What I can find the workaround is:

1. Change my server timezone to UTC, or 
2. Change my MySQL field date type from `DATE` to `DATETIME`, then it will pass the timezone info to Budibase.

Or, is there any better solution?

**Comment by ggs331** [comment]
I have just make debug on MySQL server query log, and I found that if I select `2021-04-11` on date picker(without time), then Budibase submit `2021-04-10 16:00:00.000` to MySQL which made my date incorrect.

As my timezone is China Time(UTC +8). It seems when date picker select a date(without time), then Budibase assume it's my local date time(UTC +8), then Budibase will convert it to UTC which is  -8 in my case and send to MySQL. 

So, the problem seem the Budibase always assume the selected date/time is server local date/time, but it will convert to UTC and send it out to the external database.

**Comment by aptkingston** [comment]
I think we should actually take inspiration from databases here and offer 2 unique data types to handle this:
- Datetime (date with time, with timezone support - current behaviour but time is always present)
- Date (date only, no timezome support, simple date string)

That's consistent with how certain databases handle dates, and is explicit in terms of behaviour. The other option is to make some sort of option for "timezone aware" as part of the schema for date fields. But does it make sense to have a date picker with a time picker that doesn't understand timezones? I personally don't think it does, which is why I like the idea of having 2 discrete date types, one of which always supports timestamps and is timezone aware.

Any thoughts @mjashanks?

**Comment by mjashanks** [comment]
I think it does make sense to have a date picker with time, that doesn't understand timezones.

Let's say I am walking through the factory floor, completing safety checks on machinery. I want to record the exact time that I completed this check, so I choose `now` on the picker - `26th Novemeber 2021, 14:50`

I live in California - `UTC - 7:00`. The time gets saved in my database as `26th Novemeber 2021, 21:50` (UTC).

Our company only works in one timezone.

Now - when someone looks at the data in Budibase - it's all good - reads as `26th Novemeber 2021, 14:50`

However, I have other applications reading from this database. They read the time as `26th Novemeber 2021, 21:50` - because when we developed those applications, we didn't care about timezones.

By always converting times to UTC we are forcing our decisions on other people's data. These decisions have already been made before Budibase came along. 

I think there is a use for `Date` and `DateTime`. But `with timezone` and `without timezone` is another issue.

I think we need `DateTime with timezone` and `DateTime without timezone`. 

For me `Date` and `DateTime` is a configuration of the picker. `with timezone` is a configuration of schema. But I'm not too bothered how it's implemented.

**Comment by Rory-Powell** [comment]
I don't like the idea of storing non zone aware datetimes as anything other than UTC as it's technically data loss, but given the scenario you mentioned @mjashanks, as long as all parties are operating under the same contract i.e. assumed time zone, then it should be fine. 

I don't think we need to consider storing a zoned datetime as that could compound the problem, mysql also doesn't support it on their datetime so it would have to be custom impl. Usually you wouldn't need that sort of info unless the zoned date time was for an asynchronous process e.g. a background job that I want to fire at exactly this time in this timezone respecting daylight savings. For 99% of use cases using the raw datetime with timezone conversion done either in the browser or in the api layer using location / a user profile setting is sufficient. 

For the tweaks you mentioned, maybe it could look something like this:

![Screenshot 2021-11-26 at 16 19 03](https://user-images.githubusercontent.com/8755148/143608685-a099374d-e118-4c31-af66-189c401f996c.png)

Where this is the default. When the value is set to false we would need to skip local timezone conversion in the browser (I assume this is how we do it atm, @aptkingston  correct me if I am wrong) as we have no basis to convert it to anything. 

For cases like mysql timestamp and other db engines that do auto utc conversion we would have to disable the option to change the utc setting.

**Comment by aptkingston** [comment]
I have looked at this once again. This time I used a MySQL database and simulated having the client and server either  in different time zones, or the same timezone. I also tested using a Budibase create date column (which is a `datetime`) and a manually created `timestamp` column.

 TLDR is that it works as expected (for any types that are not plain date strings.

Write up is here: [https://app.zenhub.com/files/190729906/18cb40d1-c6dd-4178-8700-36ab8c993806/download](https://app.zenhub.com/files/190729906/18cb40d1-c6dd-4178-8700-36ab8c993806/download)

Re-reading some of these comments, it looks to me like all we really need is a simple date type (without a time) that is just a date - no time, no timezone, just a date string. Databases all have equivalents of this type of data, and I think this is where the issue occurs, because we map these types to Budbase dates which are always timezone aware.

This could either be a new data type or a schema setting, but we cannot have the date picker component be in control of whether time is shown or not.

I propose:
- The current type "Date/Time" remains as a timestamp equivalent and always includes the time component. We could rename this to timestamp if that is more clear
- We add a new data type called "Date" which will never show a time picker.
- We update the SQL table import to map plain date types to the "Date" type and anything else to our "Date/Time" type, so that we account for timezones for all time related fields, and do not account for timezones for plain date strings. A separate data type is the cleanest way to ensure we handle these different cases properly.

**Comment by mjashanks** [comment]
@aptkingston its this part that is the issue:

![image](https://user-images.githubusercontent.com/3524181/160444146-31f2c40c-f7c6-4dc7-b79f-496b7f0883bc.png)

This is technically correct. Budibase is indeed doing things correctly.

However, there are scenarios where this is not what the user wants.

In your example, the data is saved as `2022-03-01 7:00:00.000`. When the user reads this field from other applications, whether it's reporting software or apps, they expect it to be `2022-03-01 15:00:00.000`. 

You can argue that the other applications are incorrect - and you would be right. However, it doesn't matter that you're right - all that matters is that it acts as the user (and their others apps) expect it to work.

We need to be able to check a box to "Ignore Timezone", to ensure that it's saves as `2022-03-01 15:00:00.000` and displays as `2022-03-01 15:00:00.000` in Budibase, regardless of timezone.

**Comment by aptkingston** [comment]
> @aptkingston its this part that is the issue:
> 
> ![image](https://user-images.githubusercontent.com/3524181/160444146-31f2c40c-f7c6-4dc7-b79f-496b7f0883bc.png)
> 
> This is technically correct. Budibase is indeed doing things correctly.
> 
> However, there are scenarios where this is not what the user wants.
> 
> In your example, the data is saved as `2022-03-01 7:00:00.000`. When the user reads this field from other applications, whether it's reporting software or apps, they expect it to be `2022-03-01 15:00:00.000`.
> 
> You can argue that the other applications are incorrect - and you would be right. However, it doesn't matter that you're right - all that matters is that it acts as the user (and their others apps) expect it to work.
> 
> We need to be able to check a box to "Ignore Timezone", to ensure that it's saves as `2022-03-01 15:00:00.000` and displays as `2022-03-01 15:00:00.000` in Budibase, regardless of timezone.

I see. I guess I just can't imagine any sort of application where data loss is desired. I think we would have to show warnings about it that what they are picking is "server time" and not their time. I don't mind adding this as long as warnings are given, because it's the sort of thing that is likely to cause more issues since it's voluntary data loss.

We also can't do this automatically at import so it would need to be a manual step after importing tables.

From reading the past issues, I also didn't get the impression that anyone had actually encountered this issue (of having client and server in different timezones), and rather were just encountering the issue of needing a raw date type rather than datetime. At least this was definitely what @ggs331 encountered. If we implement the idea of a non-timezone-aware timestamp or not, I still think a plain date type is very valuable.

**Comment by mjashanks** [comment]
Sorry, my bad - the issue exists in both scenarios. In fact its much more relevant for the same-timezone scenario

![image](https://user-images.githubusercontent.com/3524181/160459571-e06a9f95-e76d-4e66-999a-2ee4fa33d461.png)

I.e. if the users selects 1st March @ 00:00, thats exactly how it should look in the database. I agree that a Date only value would work for this scenario, unless they actually meant to pick midnight.

Think of it this way... I have an app that let people record appointment dates - i.e. date and time. Using Budibase, I record an appointment as `1st March @ 2pm`. Now, I want to write a SQL query that pulls out all afternoon appointments. My business only exists in one timezone. That becomes a difficult query to write, because I have to introduce timezone to my query.

Also, if my other apps are already ignoring timezones, then there is a mismatch in the format.

**Comment by aptkingston** [comment]
With the same timezone approach though, querying out the value again does provide the time that was picked in Budibase. At least with my MySQL experiment there, both a `timestamp` and a `datetime` correctly understand a UTC time and will return the local time when queried. (This is from a MySQL query, not budibase):
![image](https://user-images.githubusercontent.com/9075550/160552456-950277b3-18b9-4e97-9292-ea3c0839d624.png)

That's why I was saying above this issue only happens when client and server are in different timezones. It's hard to test every type of column in every type of database though, but I would assume that the major SQL databases would behave in a similar way.

**Comment by pbleic** [comment]
> Re-reading some of these comments, it looks to me like all we really need is a simple date type (without a time) that is just a date - no time, no timezone, just a date string. Databases all have equivalents of this type of data, and I think this is where the issue occurs, because we map these types to Budbase dates which are always timezone aware.
> 
> This could either be a new data type or a schema setting, but we cannot have the date picker component be in control of whether time is shown or not.
> 
> I propose:
> 
> * The current type "Date/Time" remains as a timestamp equivalent and always includes the time component. We could rename this to timestamp if that is more clear
> * We add a new data type called "Date" which will never show a time picker.
> * We update the SQL table import to map plain date types to the "Date" type and anything else to our "Date/Time" type, so that we account for timezones for all time related fields, and do not account for timezones for plain date strings. A separate data type is the cleanest way to ensure we handle these different cases properly.

Please - this is a very important function.  Creating a DATE data type would solve it.  Need to be able to search, filter, etc. on DATE without a TIME component.  Can you prioritize this?  There is no workaround.

**Comment by JayP718** [comment]
Hi,


Is there any update on this? I love Budibase but this is preventing from rolling out an application.

Best,

Jay

**Comment by JayP718** [comment]
Agreed. I also think this feature is crucial. Especially when we think about the main use case of Budibase which is too build internal apps. Only a small percentage of companies span different time zones. And when dealing with APIs that load data into a database, many times that data isn't UTC format or time-zone aware. Example) Local municipality data

**Comment by shogunpurple** [comment]
This one should now be resolved with the new "Ignore Timezones" flag introduced in https://github.com/Budibase/budibase/pull/6206. Closing
